### ISSUE TITLE
Kafka protocol portal

### ISSUE THREAD

**Issue by hairyhum** [build-trust/ockam#3963]
### Background

To provide seamless integration with existing applications using Kafka, Ockam can be used to deliver messages from Kafka client to Kafka broker (via TCP portal tunneling). 

To make it possible to encrypt Kafka messages in the broker, the portal needs to be able to parse Kafka protocol reuqests/responses, perform operations on the message payloads and reconstruct the requests/responses. This is not something TCP portal can do currently.

Also Kafka multi-server setups communicate multiple endpoints which clients need to connect to, those requests should also be parsed and reconstructed to support ockam tunneling to multiple kafka servers.

### Components required:

- KafkaPortal service
    This service shoud listen on a TCP port for kafka protocol messages, wrap them into ockam messages and forward to Ockam Kafka TunnelClient.
    This service can parse and reconstruct kafka protocol requests to encrypt/decrypt message payloads in PRODUCE and FETCH requests using encrypt/decrypt API for secure channels https://github.com/build-trust/ockam/issues/3962
- Kafka TunnelClient worker
    This worker should accept wrapped kafka requests and send them to the kafka broker

### Wrapped kafka request/response format:

```
payload: data # kafka request binary
metadata: {
  server: string # ""
}
```

Where server indicates the broker for multi-broker messaging.
For now while we don't use multiple broker mapping we can either have it always empty string,
or make the server field optional.

### Message flow

Kafka protocol requests always originate from the application kafka client.
Kafka client is configured to connect to the KafkaPortal server as if it was kafka broker.
KafkaPortal is configured with an Ockam Route to the kafka TunnelClient
Kafka TunnelClient is configured to connect to the actual kafka broker

![Kafka end-to-end - KafkaPortal message flow](https://user-images.githubusercontent.com/784180/209224083-1b243c1e-b22e-4edc-abc9-50f7c9c9510a.jpg)

**Comment by hairyhum** [comment]
### Open question

How should KafkaTunnelClient be created?
 - KafkaPortal on connection from application client can call some spawner to create a new client
 - There could be a single client on a node with some static address
 - There should be a pool of connections

This depends on the way SASL authentication mechanisms work in the particular kafka connection, probably options 2 and 3 would be easier to implement, but option 1 might work better with SASL authentication.

**Comment by hairyhum** [comment]
Portal worker behaves similarly to TCP inlet while Tunnel Client is similar to TCP outlet. They can use similar handshake protocol to set up connections.

**Comment by hairyhum** [comment]
A bit more context on the implementation of connection between protocol portal and kafka:

We can set up plain TCP transports to connect kafka client and kafka broker. There is enough functionality to support that. Application client would be connecting to the inlet and outlet would be connecting to kafka broker.

Then we can inject a separate worker to intercept tunnel messages and do something with them. This way we don't have to implement a separate inlet-outlet, just this interceptor.

Interceptor can selectively parse kafka protocol messages and filter/modify or react to them.

![Kafka end-to-end - Kafka portal](https://user-images.githubusercontent.com/784180/212131720-fa1dac7e-bad7-4ddd-8114-c99b7dc3e593.jpg)

**Comment by hairyhum** [comment]
https://kafka.apache.org/protocol.html

**Comment by hairyhum** [comment]
## About kafka protocol request header versions

Kafka protocol defines 3 versions for request header and 2 versions for response header.
Request header version is derived from the request api version.
Response header version is derived from the request api version (accessed via correlation ID)
It's not clearly defined in the docs which api versions correspond to which request/response header versions.

We can use code from https://github.com/apache/kafka to generate the code for
`org.apache.kafka.common.message.ApiMessageType`
using `./gradlew processMessages processTestMessages`

The code will be in `clients/src/generated/java/org/apache/kafka/common/message/ApiMessageType.java`

The functions we need are `requestHeaderVersion` and `responseHeaderVersion`

From there as we're interested in PRODUCE, FETCH and METADATA requests:

Request header versions:
```
case 0: // Produce
    if (_version >= 9) {
        return (short) 2;
    } else {
        return (short) 1;
    }
case 1: // Fetch
    if (_version >= 12) {
        return (short) 2;
    } else {
        return (short) 1;
    }
case 3: // Metadata
    if (_version >= 9) {
        return (short) 2;
    } else {
        return (short) 1;
    }
```

PRODUCE:
| api version | request header version |
|-------------|------------------------|
| 0-8 | 1 |
| 9 | 2 |

FETCH
| api version | request header version |
|-------------|------------------------|
| 0-11 | 1 |
| 12-13 | 2 |

METADATA:
| api version | request header version |
|-------------|------------------------|
| 0-8 | 1 |
| 9-12 | 2 |



Response header versions:
```
case 0: // Produce
    if (_version >= 9) {
        return (short) 1;
    } else {
        return (short) 0;
    }
case 1: // Fetch
    if (_version >= 12) {
        return (short) 1;
    } else {
        return (short) 0;
    }
case 3: // Metadata
    if (_version >= 9) {
        return (short) 1;
    } else {
        return (short) 0;
    }
```

PRODUCE:
| api version | response header version |
|-------------|------------------------|
| 0-8 | 0 |
| 9 | 1 |

FETCH
| api version | response header version |
|-------------|------------------------|
| 0-11 | 0 |
| 12-13 | 1 |

METADATA:
| api version | response header version |
|-------------|------------------------|
| 0-8 | 0 |
| 9-12 | 1 |

We can come back to this file if we need to intercept more API requests or versions.

@davide-baldo FYI

**Comment by hairyhum** [comment]
Addressed in #4155
